### PR TITLE
Add 'body' field to service's 'check' stanza

### PR DESCRIPTION
**Pull request by nick96**
This adds a `body` field to the service `check` stanza.

As described in https://github.com/hashicorp/nomad/issues/10084 Consul has this field so it's just a matter of parsing it out of the job config and plumbing it through.

Currently there is no validation that the `method` and `body` make sense together (e.g.`method` can be "GET" and the `body` field can exist). I couldn't find this validation in the Consul codebase either.

I'm also still having a look around the code to see where different bits are tested.

# New `jobspec` Entry Checklist

## Code

* [x] Consider similar features in Consul, Kubernetes, and other tools. Is there prior art we should match? Terminology, structure, etc? (adding something already in Consul)
* [x] Add structs/fields to `api/` package
  * `api/` structs usually have Canonicalize and Copy methods
  * New fields should be added to existing Canonicalize, Copy methods
  * Test the structs/fields via methods mentioned above
* [x] Add structs/fields to `nomad/structs` package
  * `structs/` structs usually have Copy, Equals, and Validate methods
  * Validation happens in this package and _must_ be implemented
  * Note that analogous struct field names should match with `api/` package
  * Test the structs/fields via methods mentioned above
  * Implement and test other logical methods
* [x] Add conversion between `api/` and `nomad/structs` in `command/agent/job_endpoint.go`
  * Add test for conversion
* [x] Implement diff logic for new structs/fields in `nomad/structs/diff.go` - This seems to just be handled by the generic diff function because `body` is a primitive
  * Note that fields must be listed in alphabetical order in `FieldDiff` slices in `nomad/structs/diff_test.go`
  * Add test for diff of new structs/fields
* [x] Add change detection for new structs/feilds in `scheduler/util.go/tasksUpdated` - I think this is already handle in the equality check
  * Might be covered by `.Equals` but might not be, check.
  * Should return true if the task must be replaced as a result of the change.

## HCL1 (deprecated)

New jobspec entries should only be added to `jobspec2`. It makes use of HCL2
and the `api` package for automatic parsing. Before, additional parsing was
required in the original `jobspec` package.

* [x] ~~Parse in `jobspec/parse.go`~~ (HCL1 only)
* [x] ~~Test in `jobspec/parse_test.go` (preferably with a
  `jobspec/text-fixtures/<feature>.hcl` test file)~~ (HCL1 only)

## Docs

* [x] Changelog
* [x] Jobspec entry https://www.nomadproject.io/docs/job-specification/index.html
* [ ] ~Jobspec sidebar entry https://github.com/hashicorp/nomad/blob/main/website/data/docs-navigation.js~
* [x] Job JSON API entry https://www.nomadproject.io/api/json-jobs.html
* [ ] Sample Response output in API https://www.nomadproject.io/api/jobs.html
* [x] Consider if it needs a guide https://www.nomadproject.io/guides/index.html

### CODE REVIEW

**Comment by nick96** [comment]
I've manually tested this using this job: https://gist.github.com/nick96/02b29844856de7ecafa9e1c2e8bc7c09

The CLI returns the expected health checks, including the `Body` field
``` shell
$ ./bin/nomad inspect whoami | jq '.Job.TaskGroups[0].Services[0].Checks[]'
{
  "AddressMode": "",
  "Args": null,
  "Body": "{}",
  "CheckRestart": null,
  "Command": "",
  "Expose": false,
  "FailuresBeforeCritical": 0,
  "GRPCService": "",
  "GRPCUseTLS": false,
  "Header": null,
  "Id": "",
  "InitialStatus": "",
  "Interval": 5000000000,
  "Method": "POST",
  "Name": "service: \"whoami\" check",
  "OnUpdate": "require_healthy",
  "Path": "/",
  "PortLabel": "",
  "Protocol": "",
  "SuccessBeforePassing": 0,
  "TLSSkipVerify": false,
  "TaskName": "",
  "Timeout": 5000000000,
  "Type": "http"
}
{
  "AddressMode": "",
  "Args": null,
  "Body": "",
  "CheckRestart": null,
  "Command": "",
  "Expose": false,
  "FailuresBeforeCritical": 0,
  "GRPCService": "",
  "GRPCUseTLS": false,
  "Header": null,
  "Id": "",
  "InitialStatus": "",
  "Interval": 5000000000,
  "Method": "GET",
  "Name": "service: \"whoami\" check",
  "OnUpdate": "require_healthy",
  "Path": "/health",
  "PortLabel": "",
  "Protocol": "",
  "SuccessBeforePassing": 0,
  "TLSSkipVerify": false,
  "TaskName": "",
  "Timeout": 5000000000,
  "Type": "http"
}
```

API result is the same, except it doesn't contain the the `Id` field:
``` shell
$ curl http://localhost:4646/v1/job/whoami | jq -S '.TaskGroups[0].Services[0].Checks[]'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  3484    0  3484    0     0  1134k      0 --:--:-- --:--:-- --:--:-- 1134k
{
  "AddressMode": "",
  "Args": null,
  "Body": "{}",
  "CheckRestart": null,
  "Command": "",
  "Expose": false,
  "FailuresBeforeCritical": 0,
  "GRPCService": "",
  "GRPCUseTLS": false,
  "Header": null,
  "InitialStatus": "",
  "Interval": 5000000000,
  "Method": "POST",
  "Name": "service: \"whoami\" check",
  "OnUpdate": "require_healthy",
  "Path": "/",
  "PortLabel": "",
  "Protocol": "",
  "SuccessBeforePassing": 0,
  "TLSSkipVerify": false,
  "TaskName": "",
  "Timeout": 5000000000,
  "Type": "http"
}
{
  "AddressMode": "",
  "Args": null,
  "Body": "",
  "CheckRestart": null,
  "Command": "",
  "Expose": false,
  "FailuresBeforeCritical": 0,
  "GRPCService": "",
  "GRPCUseTLS": false,
  "Header": null,
  "InitialStatus": "",
  "Interval": 5000000000,
  "Method": "GET",
  "Name": "service: \"whoami\" check",
  "OnUpdate": "require_healthy",
  "Path": "/health",
  "PortLabel": "",
  "Protocol": "",
  "SuccessBeforePassing": 0,
  "TLSSkipVerify": false,
  "TaskName": "",
  "Timeout": 5000000000,
  "Type": "http"
}
```
The health check is successfully registered in Consul:

Screen Shot 2021-03-19 at 12 30 47 am

**Comment by nick96** [comment]
Just playing around with this if I change the `body` and run `./bin/nomad run whoami.hcl` again it doesn't update the `body` field but it works for all other fields.

File: nomad/structs/diff_test.go [hunk 1]
```diff
@@ -2805,6 +2805,12 @@ func TestTaskGroupDiff(t *testing.T) {
 										Old:  "",
 										New:  "",
 									},
+									{
+										Type: DiffTypeNone,
+										Name: "Body",
+										Old:  "",
```
Comment (tgross) [inline nomad/structs/diff_test.go:2811]: While this _should_ be trivial and rightfully not worth testing, but unfortunately it's really easy for us to break diffing. 😊 So let's add a `Body` to this test.

File: CHANGELOG.md [hunk 2]
```diff
@@ -14,6 +14,7 @@ IMPROVEMENTS:
  * consul/connect: Enable setting `local_bind_address` field on connect upstreams [[GH-6248](https://github.com/hashicorp/nomad/issues/6248)]
  * driver/docker: Added support for optional extra container labels. [[GH-9885](https://github.com/hashicorp/nomad/issues/9885)]
  * driver/docker: Added support for configuring default logger behavior in the client configuration. [[GH-10156](https://github.com/hashicorp/nomad/issues/10156)]
+ * client: Allow setting `body` field on service/check Consul health checks.
```
Comment (tgross) [inline CHANGELOG.md:17]:
```suggestion
 * client/service: Allow setting `body` field on service/check Consul health checks.
```

File: jobspec/test-fixtures/tg-service-check-body.hcl [hunk 3]
```diff
@@ -0,0 +1,28 @@
+job "group_service_check_script" {
```
Comment (tgross) [inline jobspec/test-fixtures/tg-service-check-body.hcl:1]: We're not adding any new fields to HCL1, let's move this into the `jobspec2` test fixtures.

File: jobspec/parse_service.go [hunk 4]
```diff
@@ -852,6 +852,7 @@ func parseChecks(service *api.Service, checkObjs *ast.ObjectList) error {
 			"tls_skip_verify",
 			"header",
 			"method",
+			"body",
```
Comment (tgross) [inline jobspec/parse_service.go:855]: We aren't adding any new fields to HCL1 parsing. So let's remove this.

**Review by tgross — CHANGES_REQUESTED** [review]
Hi @nick96! This is looking great so far!

File: nomad/structs/services.go [hunk 5]
```diff
@@ -61,6 +61,7 @@ type ServiceCheck struct {
 	TaskName               string              // What task to execute this check in
 	SuccessBeforePassing   int                 // Number of consecutive successes required before considered healthy
 	FailuresBeforeCritical int                 // Number of consecutive failures required before considered unhealthy
+	Body                   string              // Body to use in HTTP check
```
[the remark this answers is not in the document]
Comment (nick96) [reply nomad/structs/services.go:64]: Awesome. Thank you! I've added it

File: nomad/structs/diff_test.go (hunk 1 above)
Comment (nick96) [reply nomad/structs/diff_test.go]: No worries! I've updated the test to check for this now

File: website/content/api-docs/json-jobs.mdx [hunk 6]
```diff
@@ -489,28 +489,30 @@ The `Task` object supports the following keys:
   - `Header`: Headers for HTTP checks. Should be an object where the values are an
     array of values. Headers will be written once for each value.
 
-    - `Interval`: This indicates the frequency of the health checks that
```
[the remark this answers is not in the document]
Comment (nick96) [reply website/content/api-docs/json-jobs.mdx:492]: Thanks. I've fixed it now

**Comment by nick96** [comment]
@tgross I think I've fixed all your comments. PTAL when you get the chance

**Comment by tgross** [comment]
Hi @nick96 sorry about the delay on this review. I just wrote the following job spec file and tested it end-to-end and it looks good:

<details><summary>echo server jobspec</summary>

```hcl
job "example" {
  datacenters = ["dc1"]

  group "web" {

    network {
      mode = "bridge"
      port "www" {
        to = 8000
      }
    }

    service {
      port = "www"

      check {
        type = "http"
        port = "www"
        method = "POST"
        path = "/"
        # body = "check, please"
        body = "make sure updates work, please"
        interval = "10s"
        timeout = "2s"
      }
    }

    task "httpd" {
      driver = "docker"

      config {
        image   = "python:3.8-alpine"
        command = "python"
        args    = ["/local/job.py"]
        ports   = ["www"]
      }

      template {
        data        = <<EOT
import http.server

class Handler(http.server.BaseHTTPRequestHandler):

    def do_GET(self):
        self.send_response(200)
        self.send_header('Content-type', 'text/html')
        self.end_headers()

    def do_POST(self):
        content_length = int(self.headers['Content-Length'])
        post_data = self.rfile.read(content_length)
        post_data.decode('utf-8')
        print(post_data)
        self.send_response(200)
        self.send_header('Content-type', 'text/html')
        self.end_headers()
        self.wfile.write("{}\n".format(post_data).encode('utf-8'))

def run(server_class=http.server.HTTPServer, handler_class=Handler):
    server_address = ('', 8000)
    httpd = server_class(server_address, handler_class)
    httpd.serve_forever()

if __name__ == '__main__':
    run()

EOT
        destination = "local/job.py"
      }

      resources {
        cpu    = 128
        memory = 128
      }
    }
  }
}
```

</details>

I'm going to take one more pass over this tomorrow just for style or testing issues but I think we should be good to merge shortly. Thanks for your patience!

**Comment by nick96** [comment]
No worries! I'm away for the rest of this week so won't be able to address anything but will get onto it ASAP when I'm back.

**Review by tgross — APPROVED** [review]

**Comment by tgross** [comment]
Merged!

**Pull request merged**